include/pybind11/cast.h

```diff
@@ -980,11 +980,19 @@ struct type_caster<T, enable_if_t<std::is_arithmetic<T>::value && !is_std_char_t
     static handle cast(T src, return_value_policy /* policy */, handle /* parent */) {
         if (std::is_floating_point<T>::value) {
             return PyFloat_FromDouble((double) src);
-        } else if (sizeof(T) <= sizeof(long)) {
+        } else if (sizeof(T) <= sizeof(ssize_t)) {
+#if PY_VERSION_HEX < 0x03000000
```

Comment:
Instead of the `#ifdef` here, let's add these to `detail/common.h` like we have for various other python API calls that differ across python versions:
```C++
#define PYBIND11_LONG_FROM_SIGNED(o) PyLong_FromSize_t((size_t) o)
#define PYBIND11_LONG_FROM_UNSIGNED(o) PyLong_FromSsize_t((ssize_t) o)
```
for Python 3 and
```C++
#define PYBIND11_LONG_FROM_SIGNED(o) PyInt_FromSize_t((size_t) o)     // Returns long if needed.
#define PYBIND11_LONG_FROM_UNSIGNED(o) PyInt_FromSsize_t((ssize_t) o) // Likewise.
```
for Python 2; then this part of the `cast.h` code becomes much cleaner:
```C++
            if (std::is_signed<T>::value)
                return PYBIND11_LONG_FROM_SIGNED(o);
            else
                PYBIND11_LONG_FROM_UNSIGNED(o);
```

Reply:
Okay, I was considering something like that. I'll add when the CI build finishes.

Reply:
Done.